File: aoc-2021/packages/aoc-go/solutions/day_05.go
package solutions

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type orientation int

const (
	vertical orientation = iota + 1
	horizontal
	diagonal45
	other
)

type Point struct {
	x, y int
}

type Line struct {
	start Point
	end   Point
}

func (l Line) max() int {
	max := l.start.x
	if l.start.y > max {
		max = l.start.y
	}
	if l.end.x > max {
		max = l.end.x
	}
	if l.end.y > max {
		max = l.end.y
	}
	return max
}

func (l Line) orientation() orientation {
	if l.start.x == l.end.x {
		return vertical
	}
	if l.start.y == l.end.y {
		return horizontal
	}
	if math.Abs(float64(l.start.x-l.end.x)) == math.Abs(float64(l.start.y-l.end.y)) {
		return diagonal45
	}
	return other
}

func parsePoint(data string) Point {
	points := strings.Split(data, ",")
	x, _ := strconv.Atoi(points[0])
	y, _ := strconv.Atoi(points[1])
	return Point{x, y}
}

func parseLine(line string) Line {
	points := strings.Split(line, " -> ")
	return Line{
		start: parsePoint(points[0]),
		end:   parsePoint(points[1]),
	}
}

func printBoard(board [][]int) {
	for _, row := range board {
		for _, col := range row {
			if col == 0 {
				fmt.Print(". ")
			} else {
				fmt.Printf("%d ", col)
			}
		}
		fmt.Println("")
	}
	fmt.Println("")
}

func ShowD05() {
	lines := []Line{}
	file, _ := os.Open("../../input/day_05.in")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, parseLine(scanner.Text()))
	}

	max := 0
	for _, l := range lines {
		lineSize := l.max()
		if lineSize > max {
			max = lineSize
		}
	}

	board := make([][]int, max+1)
	for i := 0; i <= max; i++ {
		board[i] = make([]int, max+1)
	}

	count := 0
	for _, l := range lines {
		if l.orientation() != vertical && l.orientation() != horizontal {
			continue
		}
		start := int(math.Min(float64(l.start.x), float64(l.end.x)))
		end := int(math.Max(float64(l.start.x), float64(l.end.x)))
		if l.orientation() == vertical {
			start = int(math.Min(float64(l.start.y), float64(l.end.y)))
			end = int(math.Max(float64(l.start.y), float64(l.end.y)))
		}
		for start <= end {
			x := start
			y := l.start.y
			if l.orientation() == vertical {
				x = l.start.x
				y = start
			}
			board[y][x] += 1
			if board[y][x] == 2 {
				count += 1
			}
			start += 1
		}
	}

	printBoard(board)
	fmt.Printf("Part 1: %d\n", count)
	for _, l := range lines {
		if l.orientation() != diagonal45 {
			continue
		}
		m := (l.end.x - l.start.x) / (l.end.y - l.start.y)
		yMax := l.start.y
		yMin := l.end.y
		if yMax < yMin {
			yMax, yMin = yMin, yMax
		}
		col := l.start.x
		xMax := l.end.x
		if col > xMax {
			col, xMax = xMax, col
		}
		row := yMin
		if m == -1 {
			row = yMax
		}

		for col <= xMax {
			board[row][col] += 1
			if board[row][col] == 2 {
				count += 1
			}
			col += 1
			row += m
		}

	}

	printBoard(board)
	fmt.Printf("Part 2: %d\n", count)
}

File: aoc-rust/src/2021/day_05.rs
use std::cmp::{max, min};
use std::ops::{Range, RangeInclusive};
use std::str::FromStr;

#[derive(Debug)]
struct Line {
    x: (usize, usize),
    y: (usize, usize),
}

impl FromStr for Line {
    type Err = std::io::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (start, end) = s.split_once(" -> ").unwrap();
        let (x1, y1) = start.split_once(",").unwrap();
        let (x2, y2) = end.split_once(",").unwrap();
        let x1 = x1.parse::<usize>().unwrap();
        let x2 = x2.parse::<usize>().unwrap();
        let y1 = y1.parse::<usize>().unwrap();
        let y2 = y2.parse::<usize>().unwrap();
        return Ok(Line {
            x: (x1, x2),
            y: (y1, y2),
        });
    }
}

enum Orientation {
    Vertical,
    Horizontal,
    Diagonal45,
    Other,
}

impl Line {
    fn orientation(&self) -> Orientation {
        if self.x.0 == self.x.1 {
            return Orientation::Vertical;
        } else if self.y.0 == self.y.1 {
            return Orientation::Horizontal;
        } else if (self.x.0 as i32 - self.x.1 as i32).abs()
            == (self.y.0 as i32 - self.y.1 as i32).abs()
        {
            return Orientation::Diagonal45;
        }
        Orientation::Other
    }

    fn is_straight(&self) -> bool {
        match self.orientation() {
            Orientation::Vertical | Orientation::Horizontal => true,
            _ => false,
        }
    }
}

fn print_board(board: &Vec<Vec<i32>>) {
    for l in board.iter() {
        for c in l.iter() {
            print!(
                "{} ",
                if *c == 0 {
                    ".".to_string()
                } else {
                    c.to_string()
                }
            );
        }
        println!("");
    }
    println!("");
}

pub fn show() {
    let lines: Vec<Line> = include_str!("../../../input/day_05.in")
        .lines()
        .map(|x| x.parse().unwrap())
        .collect();

    // Get the size of the matrix
    let size = lines.iter().fold(0, |mut max, line| {
        let max_line = vec![line.x.0, line.x.1, line.y.0, line.y.1];
        match max_line.iter().max() {
            Some(m) => {
                if *m > max {
                    max = *m
                }
            }
            _ => {}
        }
        max
    });
    let (board, count) = lines.iter().filter(|l| l.is_straight()).fold(
        (vec![vec![0; size + 1]; size + 1], 0),
        |(mut board, mut count), line| {
            let vertical = matches!(line.orientation(), Orientation::Vertical);
            let mut start = min(line.x.0, line.x.1);
            let mut end = max(line.x.0, line.x.1);
            if vertical {
                start = min(line.y.0, line.y.1);
                end = max(line.y.0, line.y.1)
            }

            while start <= end {
                let x = if vertical { line.x.0 } else { start };
                let y = if vertical { start } else { line.y.0 };
                board[y][x] += 1;
                if board[y][x] == 2 {
                    count += 1;
                }
                start += 1;
            }
            return (board, count);
        },
    );
    //print_board(&board);
    println!("- Part 1: {}", count);

    let (board, count) = lines
        .iter()
        .filter(|l| matches!(l.orientation(), Orientation::Diagonal45))
        .fold((board, count), |(mut board, mut count), line| {
            let m: i32 = (line.x.1 as i32 - line.x.0 as i32) / (line.y.1 as i32 - line.y.0 as i32);
            let y_max = max(line.y.0, line.y.1);
            let y_min = min(line.y.0, line.y.1);
            // Always go from left to right
            let mut col = min(line.x.0, line.x.1);
            // Start from top or bottom depending on the line's orientation
            let mut row = (if m == -1 { y_max } else { y_min }) as i32;

            while col <= max(line.x.0, line.x.1) {
                board[row as usize][col] += 1;
                if board[row as usize][col] == 2 {
                    count += 1;
                }

                col += 1;
                row += m;
            }
            return (board, count);
        });
    //print_board(&board);
    println!("- Part 2: {}", count);
}
